Fetch ConfigMap in ModelCatalogContext provider and move provider to route level

YuliaKrimerman opened the pull request:
<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- https://issues.redhat.com/browse/RHOAIENG-123456 -->
Closes [RHOAIENG-18959](https://issues.redhat.com/browse/RHOAIENG-18959)

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->
Created the fetching of the configmap and verified that it's stringifies it in the way we need for all the models. 


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested on the green cluster. To test it please add /tempDetails to the model catalog url and checkout the console is printing Model Catalog Context. 
![Screenshot 2025-02-11 at 12 51 19 PM](https://github.com/user-attachments/assets/161d5cb4-0120-44e5-b53f-e981bde1e4f3)
## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->
Tests will be added once the work on the Empty/Loading state is done

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

mturley commented:
@YuliaKrimerman @ppadti  we need to remove the `<ModelCatalogContextProvider>` that is still in frontend/src/app/App.tsx, it probably makes sense for that to be in scope for this PR since data fetching is what makes that render location matter. I missed that when reviewing https://github.com/opendatahub-io/odh-dashboard/pull/3733, we have redundant context providers now (a provider has already been added to ModelCatalogCoreLoader).

@YuliaKrimerman you also have a merge conflict here

ppadti commented:
> @YuliaKrimerman @ppadti we need to remove the `<ModelCatalogContextProvider>` that is still in frontend/src/app/App.tsx, it probably makes sense for that to be in scope for this PR since data fetching is what makes that render location matter. I missed that when reviewing #3733, we have redundant context providers now (a provider has already been added to ModelCatalogCoreLoader).
> 
> Actually with the changes so far in this PR we now have that provider rendered in 3 places 😅 we can also revert the changes you have here to ModelCatalogRoutes. That'll resolve the merge conflict. The context provider rendered in ModelCatalogCoreLoader is what we want to keep - that will only be rendered when you're on one of these routes - the way this diff has it wrapped around the routes, it'll be rendered no matter which route you're on. Sorry, I could have been more clear in the issue description about that.

Sorry, I missed that...

YuliaKrimerman commented:
@mturley @ppadti Latest changes : removed the context from App, reverted changes in the ModelCatalogRoutes, added logging to show we're fetching the configmap. Here is a screenshot that we're parsing it correctly as well 
![Screenshot 2025-02-10 at 11 07 29 AM](https://github.com/user-attachments/assets/6034c8fc-25f4-4f5d-9211-70c9c2765945)

mturley commented on frontend/src/concepts/modelCatalog/context/ModelCatalogContext.tsx:
```diff
@@ -1,29 +1,32 @@
   children,
 }) => {
-  const modelCatalogSourcesMock = React.useMemo(() => [modelCatalogSourceMock({})], []);
+  const modelCatalogSources = useModelCatalogSources();
```
You can wrap this in `useMakeFetchObject`. This changes the array return value to an object with named properties, easier to deal with.

```ts
const modelCatalogSources = useMakeFetchObject(useModelCatalogSources());
```

mturley commented on frontend/src/concepts/modelCatalog/types.ts:
```diff
@@ -39,3 +42,18 @@ export type ModelCatalogSource = {
+
+// Temporary type for MVP - will be replaced with remote model catalog sources
+// See: https://issues.redhat.com/browse/RHOAISTRAT-455
+export type ModelCatalogConfigMap = K8sResourceCommon & {
+  metadata: {
+    name: string;
+    labels: DashboardLabels & {
+      'opendatahub.io/model-catalog': 'true';
+    };
+  };
+  data?: {
+    source: string;
+    models: string; // JSON string of CatalogModel[]
+  };
+};
```
You know what? I'm sorry - I think we can just delete this whole type. It's not used anywhere - I was about to suggest how we can use a type guard to have the value of `getConfigMap` use this type below, but the only thing that would do for us is protect against typos in `configMap.data?.modelCatalogSource` which you write on two lines - not worth the effort for a temporary workaround. Let's ditch this type.

mturley commented on frontend/src/pages/modelCatalog/ModelCatalogPage.tsx:
```diff
@@ -0,0 +1,16 @@
+
+const ModelCatalogPage: React.FC = () => {
+  const { modelCatalogSources } = React.useContext(ModelCatalogContext);
+  const [sources, loaded, error] = modelCatalogSources;
+
+  React.useEffect(() => {
+    // eslint-disable-next-line no-console
+    console.log('Loaded:', loaded);
+  }, [sources, loaded, error]);
```
You'll have to change how you consume this with the change to FetchStateObject - you'll just have `modelCatalogSources.data` (and `.loaded` and `.error`).

mturley commented on frontend/src/concepts/modelCatalog/useModelCatalogSources.ts:
```diff
@@ -0,0 +1,36 @@
+      return [];
+    }
+
+    const source: ModelCatalogSource = JSON.parse(configMap.data.modelCatalogSource);
+    return [source];
```
JSON.parse can throw an exception, and we don't want invalid JSON in the configmap to crash the UI. You'll need to wrap these 2 lines in a `try { } catch(e) { }`, and in the catch block just return `[]`. I think it's ok to swallow errors like this for our temporary summit implementation.

mturley reviewed (changes requested):
Looking really good, just a few minor things left.

mturley commented on frontend/src/pages/modelCatalog/ModelCatalogPage.tsx:
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import { ModelCatalogContext } from '~/concepts/modelCatalog/context/ModelCatalogContext';
+
+const ModelCatalogPage: React.FC = () => {
```
This page actually isn't rendered by any route - I think the page component was already created under `frontend/src/pages/modelCatalog/screens/ModelCatalog.tsx`, you can delete this file and use that one to test taking your data from context instead

mturley reviewed (approved):
LGTM! Nice work! Tested on ui-green cluster using this route: http://localhost:4010/modelCatalog/Red%20Hat/rhelai1/granite-8b-code-instruct/1%252E3%252E0
(that is valid with the configmap as it exists there currently)
